client/pages/auth-form.jsx

```diff
@@ -31,42 +38,75 @@ export default class AuthForm extends React.Component {
   handleSubmit(event) {
     event.preventDefault();
     const { username, password } = this.state;
+    const { route, handleSignIn } = this.context; // eslint-disable-line
+    const { path } = route;
 
     let usernameTooShort = false;
     let passwordTooShort = false;
+    let usernameTaken = false;
+    let invalidLogin = false;
     if (username.length < 4) {
       usernameTooShort = true;
     }
     if (password.length < 6) {
       passwordTooShort = true;
     }
     if (usernameTooShort || passwordTooShort) {
-      this.setState({ usernameTooShort, passwordTooShort });
+      this.setState({ usernameTooShort, passwordTooShort, usernameTaken, invalidLogin });
       return;
     }
 
-    const body = { username, password };
-    const req = {
-      method: 'POST',
-      headers: {
-        'Content-Type': 'application/json'
-      },
-      body: JSON.stringify(body)
-    };
-    fetch('/api/auth/sign-up', req)
-      .then(res => {
-        if (res.status === 204) {
-          this.setState({ usernameTooShort, usernameTaken: true, passwordTooShort });
-          return;
-        }
-        return res.json();
-      })
-      .then(result => {
-        if (!result) {
-          return;
-        }
-        this.setState({ username: '', password: '', usernameTooShort, usernameTaken: false, passwordTooShort });
-      });
+    if (path === 'sign-up') {
+      const body = { username, password };
+      const req = {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json'
+        },
+        body: JSON.stringify(body)
+      };
+      fetch('/api/auth/sign-up', req)
+        .then(res => {
+          if (res.status === 204) {
+            usernameTaken = true;
+            this.setState({ usernameTooShort, usernameTaken, invalidLogin, passwordTooShort });
+            return;
+          }
+          return res.json();
+        })
+        .then(result => {
+          if (!result) {
+            return;
+          }
+          usernameTaken = false;
+          this.setState({ username: '', password: '', usernameTooShort, usernameTaken, invalidLogin, passwordTooShort });
+        });
+    } else if (path === 'sign-in') {
+      const body = { username, password };
+      const req = {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json'
+        },
+        body: JSON.stringify(body)
+      };
+      fetch('/api/auth/sign-in', req)
+        .then(res => {
+          if (res.status === 401) {
+            invalidLogin = true;
+            this.setState({ usernameTooShort, usernameTaken, invalidLogin, passwordTooShort });
+            return;
+          }
+          return res.json();
+        })
+        .then(result => {
+          if (!result) {
+            return;
+          }
```

Comment:
this would be much cleaner if it were separated into two different methods that you choose from within `handleSubmit()`